README
NAME
    RT::Extension::AddAttachmentsFromTransactions - Add Attachments From
    Transactions

DESCRIPTION
    With this plugin you can attach attachments from previous transactions
    to a reply or comment.

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        If you are using RT 4.2 or greater, add this line:

            Plugin('RT::Extension::AddAttachmentsFromTransactions');

        For RT 4.0, add this line:

            Set(@Plugins, qw(RT::Extension::AddAttachmentsFromTransactions));

        or add RT::Extension::AddAttachmentsFromTransactions to your
        existing @Plugins line.

    Clear your mason cache
            rm -rf /opt/rt4/var/mason_data/obj/*

    Restart your webserver

AUTHOR
    Christian Loos <cloos@netsandbox.de>

LICENSE AND COPYRIGHT
    This software is Copyright (C) 2012-2014, Christian Loos.

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991

SEE ALSO
    <http://bestpractical.com/rt/>
    <http://www.gossamer-threads.com/lists/rt/users/107976>
    <https://github.com/bestpractical/rt/tree/4.4/attach-from-transactions>

THANKS
    Thanks to BÁLINT Bekény for contributing the code from his
    implementation.

    Also Thanks to Best Practical Solutions who are working on this feature
    for RT 4.4 on the '4.4/attach-from-transactions' branch where I've
    borrowed some code for this extension.



Changes
Revision history for RT::Extension::AddAttachmentsFromTransactions

2.03 2014-09-10
 - update license from Perl to GPLv2 to match RT license
 - make tests author test
 - update Module::Install

2.02 2014-03-12
 - fix header display for transactions in RT 4.2

2.01 2013-12-10
 - use rolled up widget for RT version 4.0.18 and above

2.00 2013-10-29
 - RT 4.2 compatibility

1.00 2013-09-05
 - add basic tests
 - Prefix headers with X for RFC822 compliance 
 - add missing dependency for List::MoreUtils
 - Ensure consistent escaping of filenames in attachment URIs

0.01 2013-04-04
 - initial release


lib/RT/Action/SendEmail_Vendor.pm
use strict;
no warnings qw(redefine);

package RT::Action::SendEmail;

# NOTE: AddAttachments method from 4.0-trunk
sub AddAttachments {
    my $self = shift;

    my $MIMEObj = $self->TemplateObj->MIMEObj;

    $MIMEObj->head->delete('RT-Attach-Message');

    my $attachments = RT::Attachments->new( RT->SystemUser );
    $attachments->Limit(
        FIELD => 'TransactionId',
        VALUE => $self->TransactionObj->Id
    );

    # Don't attach anything blank
    $attachments->LimitNotEmpty;
    $attachments->OrderBy( FIELD => 'id' );

    # We want to make sure that we don't include the attachment that's
    # being used as the "Content" of this message" unless that attachment's
    # content type is not like text/...
    my $transaction_content_obj = $self->TransactionObj->ContentObj;

    if (   $transaction_content_obj
        && $transaction_content_obj->ContentType =~ m{text/}i )
    {
        # If this was part of a multipart/alternative, skip all of the kids
        my $parent = $transaction_content_obj->ParentObj;
        if ($parent and $parent->Id and $parent->ContentType eq "multipart/alternative") {
            $attachments->Limit(
                ENTRYAGGREGATOR => 'AND',
                FIELD           => 'parent',
                OPERATOR        => '!=',
                VALUE           => $parent->Id,
            );
        } else {
            $attachments->Limit(
                ENTRYAGGREGATOR => 'AND',
                FIELD           => 'id',
                OPERATOR        => '!=',
                VALUE           => $transaction_content_obj->Id,
            );
        }
    }

    # attach any of this transaction's attachments
    my $seen_attachment = 0;
    while ( my $attach = $attachments->Next ) {
        if ( !$seen_attachment ) {
            $MIMEObj->make_multipart( 'mixed', Force => 1 );
            $seen_attachment = 1;
        }
        $self->AddAttachment($attach);
    }

    # NOTE: call the new method AddAttachmentsFromSession
    $self->AddAttachmentsFromSession();

    # NOTE: call the new method AddAttachmentsFromHeaders
    $self->AddAttachmentsFromHeaders();

}

# NOTE: new AddAttachmentsFromSession method
sub AddAttachmentsFromSession {
    my $self = shift;
    my $email = $self->TemplateObj->MIMEObj;

    # move the Attachment id's from session to the RT-Attach header
    for my $id ( @{ $HTML::Mason::Commands::session{'AttachExisting'} } ) {
        $email->head->add('RT-Attach' => $id);
    }
}

# NOTE: new AddAttachmentsFromHeaders method from 4.2/attach-from-transactions branch
sub AddAttachmentsFromHeaders {
    my $self  = shift;
    my $orig  = $self->TransactionObj->Attachments->First;
    my $email = $self->TemplateObj->MIMEObj;

    use List::MoreUtils qw(uniq);

    # Add the RT-Attach headers from the transaction to the email
    if ($orig and $orig->GetHeader('RT-Attach')) {
        for my $id ($orig->ContentAsMIME(Children => 0)->head->get_all('RT-Attach')) {
            $email->head->add('RT-Attach' => $id);
        }
    }

    # Take all RT-Attach headers and add the attachments to the outgoing mail
    for my $id (uniq $email->head->get_all('RT-Attach')) {
        $id =~ s/(?:^\s*|\s*$)//g;

        my $attach = RT::Attachment->new( $self->TransactionObj->CreatorObj );
        $attach->Load($id);
        next unless $attach->Id
                and $attach->TransactionObj->CurrentUserCanSee;

        if ( !$email->is_multipart ) {
            $email->make_multipart( 'mixed', Force => 1 );
        }
        $self->AddAttachment($attach, $email);
    }
}

1;


lib/RT/Extension/AddAttachmentsFromTransactions.pm
package RT::Extension::AddAttachmentsFromTransactions;

use 5.008003;
use strict;
use warnings;

our $VERSION = '2.03';

{
    require RT::Ticket;
    my $orig = RT::Ticket->can('_RecordNote');
    no warnings 'redefine';
    *RT::Ticket::_RecordNote = sub {
        my $self = shift;
        my %args = @_;

        # We can't do anything if we don't have an MIMEObj
        # so let the original method handle it
        return $orig->($self, %args) unless $args{'MIMEObj'};

        # move the Attachment id's from session to the X-RT-Attach header
        for my $id ( @{ $HTML::Mason::Commands::session{'AttachExisting'} } ) {
            $args{'MIMEObj'}->head->add( 'X-RT-Attach' => $id );
        }

        # cleanup session
        delete $HTML::Mason::Commands::session{'AttachExisting'};

        return $orig->($self, %args);
    };
}

{
    require RT::Action::SendEmail;
    my $orig = RT::Action::SendEmail->can('AddAttachments');
    no warnings 'redefine';
    *RT::Action::SendEmail::AddAttachments = sub {
        my $self = shift;

        $orig->($self, @_);

        $self->AddAttachmentsFromHeaders();
    };
}

{
    package RT::Action::SendEmail;

    sub AddAttachmentsFromHeaders {
        my $self  = shift;
        my $orig  = $self->TransactionObj->Attachments->First;
        my $email = $self->TemplateObj->MIMEObj;

        use List::MoreUtils qw(uniq);

        # Add the X-RT-Attach headers from the transaction to the email
        if ($orig and $orig->GetHeader('X-RT-Attach')) {
            for my $id ($orig->ContentAsMIME(Children => 0)->head->get_all('X-RT-Attach')) {
                $email->head->add('X-RT-Attach' => $id);
            }
        }

        # Take all X-RT-Attach headers and add the attachments to the outgoing mail
        for my $id (uniq $email->head->get_all('X-RT-Attach')) {
            $id =~ s/(?:^\s*|\s*$)//g;

            my $attach = RT::Attachment->new( $self->TransactionObj->CreatorObj );
            $attach->Load($id);
            next unless $attach->Id
                and $attach->TransactionObj->CurrentUserCanSee;

            if ( !$email->is_multipart ) {
                $email->make_multipart( 'mixed', Force => 1 );
            }
            $self->AddAttachment($attach, $email);
        }
    }
}

{
    package RT::Attachment;

    unless ( RT::Attachment->can('FriendlyContentLength') ) {
        *FriendlyContentLength = sub {
            my $self = shift;
            my $size = $self->ContentLength;
            return '' unless $size;

            my $kb = int($size/102.4) / 10;
            my $units = RT->Config->Get('AttachmentUnits');

            if (!defined($units)) {
                if ($size > 1024) {
                    $size = $kb . "k";
                } else {
                    $size = $size . "b";
                }
            } elsif ($units eq 'k') {
                $size = $kb . "k";
            } else {
                $size = $size . "b";
            }

            return $size;
        }
    }
}

=encoding utf8

=head1 NAME

RT::Extension::AddAttachmentsFromTransactions - Add Attachments From Transactions

=head1 DESCRIPTION

With this plugin you can attach attachments from previous transactions to a
reply or comment.

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt4/etc/RT_SiteConfig.pm>

If you are using RT 4.2 or greater, add this line:

    Plugin('RT::Extension::AddAttachmentsFromTransactions');

For RT 4.0, add this line:

    Set(@Plugins, qw(RT::Extension::AddAttachmentsFromTransactions));

or add C<RT::Extension::AddAttachmentsFromTransactions> to your existing C<@Plugins> line.

=item Clear your mason cache

    rm -rf /opt/rt4/var/mason_data/obj/*

=item Restart your webserver

=back

=head1 AUTHOR

Christian Loos <cloos@netsandbox.de>

=head1 LICENSE AND COPYRIGHT

This software is Copyright (C) 2012-2014, Christian Loos.

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=head1 SEE ALSO

=over

=item L<http://bestpractical.com/rt/>

=item L<http://www.gossamer-threads.com/lists/rt/users/107976>

=item L<https://github.com/bestpractical/rt/tree/4.4/attach-from-transactions>

=back

=head1 THANKS

Thanks to BÁLINT Bekény for contributing the code from his implementation.

Also Thanks to Best Practical Solutions who are working on this feature for
RT 4.4 on the '4.4/attach-from-transactions' branch where I've borrowed some
code for this extension.

=cut

1;
